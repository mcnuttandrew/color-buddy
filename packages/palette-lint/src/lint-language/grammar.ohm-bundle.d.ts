// AUTOGENERATED FILE
// This file was generated from grammar.ohm by `ohm generateBundles`.

import {
  BaseActionDict,
  Grammar,
  IterationNode,
  Namespace,
  Node,
  NonterminalNode,
  Semantics,
  TerminalNode
} from 'ohm-js';

declare const ns: {
};
export default ns;

