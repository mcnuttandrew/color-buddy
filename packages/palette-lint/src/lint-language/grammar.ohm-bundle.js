'use strict';const {makeRecipe}=require('ohm-js');const result={};module.exports=result;