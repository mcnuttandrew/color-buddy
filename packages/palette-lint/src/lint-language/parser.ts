import * as ohm from "ohm-js";
const gram = ohm.grammar(String.raw`
LintLanguage {
  Program = Exp
  Exp 
  	= "NOT" paren<Exp> -- not
    | CompareOp
    | QuantExp
    | QuantExpWithWhere
    | paren<Exp>
    | BoolOp
    | bool
  
  CompareOp = paren<Exp> CompareOpType paren<Exp>
  CompareOpType = "AND" | "OR"
  
  // missing where stuff  
  QuantExp = ("ALL" | "EXIST") paren<ListOf<Variable, ",">> "IN" paren<Value> "SUCH THAT" paren<Exp> 
  QuantExpWithWhere = ("ALL" | "EXIST") paren<ListOf<Variable, ",">> "IN" paren<Value> QuantWhere "SUCH THAT" paren<Exp> 
  QuantWhere = "WHERE" paren<Exp>
  Value 
  	= 
    | Array
    | ValOp
    | ColorOp
    | ArrayOp
    | color
    | number
    | Variable
  
  Variable =  Array | IndexShortHand | ident
  IndexShortHand = "index(" ident ")"
    
  BoolOp = BoolOpBinary | BoolOpFunc
  BoolOpBinary = Value BinOps Value
  BinOps = "<" | ">" | "==" | "!="
  BoolOpFunc = BoolOpFuncOps "(" Value "," Value ("," Value)? ")"
  BoolOpFuncOps = "similar" | "isTag" 
   
  ValOp = 
   | "absDiff" "(" Value "," Value ")" -- absDiff
   | Value ValsOps Value -- normal

  ValsOps = "%" | "*" | "+" | "-" 
   
  ArrayOp = ArraySimpleOp | ArrayHoF
  ArraySimpleOp = ArrayOpType "(" Value ")" 
  ArrayOpType = "count" | "sum" | "min" | "max" | "std" | "mean" | "first" | "last" | "extent" | "speed" | "reverse"
  ArrayHoF = ArrayHofType "(" Value "," Lambda ")" 
  ArrayHofType = "sort" | "filter" | "map"
 
  
  Lambda = ident "=>" Value
 
  ColorOp = 
  	| ColorOpOneFuncs  
    | ToSpaceShortHand -- shorthand
    | ColorOpTwo
  ColorOpOneFuncs = Func1<ColorOpOneFuncsOps>
  ColorOpOneFuncsOps = 
  	"cvdSim" | "name" | "inGamut" | ToSpaceShortHand
  ColorOpTwo = Func2<ColorOpTwoOps>
  ColorOpTwoOps = "dist" | "deltaE" | "contrast"
    
  ToSpaceShortHand = (letter letter letter "." letter) "(" Value ")"
 
  Func1<name> = name "(" Value ("," Value)? ")"
  Func2<name> = name "(" Value "," Value ("," Value)? ")"
 

  
  ident  (an identifier)
    =  (alnum | "_")+

  number  (a number)
    = digit* "." digit+  -- fract
    | digit+             -- whole
    
  bool = "true" | "false"
    
  color = "#"alnum alnum alnum alnum alnum alnum
    
  Array = "[" ListOf<Value, ","> "]"
  lParen = ("(")?
  rParen = (")")?
  paren<x> = lParen x rParen
 }`);

const semantics = gram.createSemantics();

const actionDict: Parameters<(typeof semantics)["addOperation"]>[1] = {
  Program(exp) {
    console.log("Program");
    return exp.buildAst();
  },
  Exp(exp) {
    console.log("Exp");
    return exp.buildAst();
  },
  Exp_not(_not, exp) {
    console.log("Exp_not");
    return { not: exp.buildAst() };
  },
  CompareOp(left, op, right) {
    console.log("CompareOp????", this.sourceString);
    const leftChild = left.buildAst();
    const rightChild = right.buildAst();
    const name = op.sourceString.toLowerCase();
    let children = [leftChild];
    // if all of the children are the same type, flatten
    if (name === "or" && rightChild.or) {
      children = [leftChild, ...rightChild.or];
    } else if (name === "and" && rightChild.and) {
      children = [leftChild, ...rightChild.and];
    } else {
      children.push(rightChild);
    }

    return { [name]: children };
  },
  QuantExp(quantType, vars, _, inVal, where, predicate) {
    console.log(" QuantExp");
    let varb = vars.buildAst();
    const varbKey = vars.children.length === 3 ? "varb" : "varbs";
    console.log("lmmlmlm", vars.sourceString, vars.numChildren);
    if (varbKey === "varb") {
      varb = varb[0];
    }
    const inKey = inVal.buildAst();
    console.log("inKey", inKey);
    const name = quantType.sourceString.toLowerCase();
    const out = {
      //   type: "QuantExp",
      [name]: {
        [varbKey]: varb,
        in: inKey,
        predicate: predicate.buildAst(),
      },
    };
    const whereAST = where.buildAst();
    if (whereAST) {
      out[name].where = whereAST;
    }
    return out;
  },
  QuantExpWithWhere(quantType, vars, _, val, where, _st, predicate) {
    return (actionDict as any).QuantExp(
      quantType,
      vars,
      _,
      val,
      where,
      predicate
    );
  },
  QuantWhere(_where, exp) {
    console.log("QuantWhere");
    return exp.buildAst();
  },
  BoolOpBinary(left, op, right) {
    console.log("BoolOpBinary");
    return {
      //   type: "BoolOpBinary",
      [op.sourceString]: {
        left: left.buildAst(),
        right: right.buildAst(),
      },
    };
  },
  BoolOpFunc(name, _lParen, term1, _comma, term2, _comma2, term3, _rParen) {
    console.log("BoolOpFunc", this.sourceString);
    if (name.sourceString === "similar") {
      return {
        similar: {
          left: term1.buildAst(),
          right: term2.buildAst(),
          threshold: term3.buildAst()[0],
        },
      };
    }
    return {
      [name.sourceString]: term1.buildAst(),
      value: term2.buildAst(),
    };
  },
  ToSpaceShortHand(a, b, c, _dot, channel, _lParen, value, _r) {
    console.log("ToSpaceShortHand");
    const spaceName = [a, b, c].map((x) => x.sourceString).join("");
    return { [`${spaceName}.${channel.sourceString}`]: value.buildAst() };
  },

  IndexShortHand(start, name, end) {
    console.log("IndexShortHand");
    return [start, name, end].map((x) => x.sourceString).join("");
  },
  Array(_a, b, _c) {
    console.log("Array", b.sourceString, b.numChildren);
    return b.buildAst();
  },
  ArrayOp(a) {
    console.log("ArrayOp");
    return a.buildAst();
  },
  ArraySimpleOp(ArrayOpType, _x, Value, _y) {
    console.log("ArraySimpleOp");
    return { [ArrayOpType.sourceString]: Value.buildAst() };
  },
  ArrayHoF(name, _lParen, term1, _comma, lambda, _rParen) {
    console.log("ArrayHoF");
    const lambdaAST = lambda.buildAst();
    return {
      [name.sourceString]: term1.buildAst(),
      func: lambdaAST.value,
      varb: lambdaAST.name,
    };
  },
  Lambda(name, _arrow, value) {
    return {
      //   type: "Lambda",
      name: name.sourceString,
      value: value.buildAst(),
    };
  },
  Value(a) {
    console.log("Value", a.ctorName);
    if (a.ctorName === "color") return a.sourceString;
    if (a.ctorName === "number") return Number(a.sourceString);
    if (a.ctorName === "Array") return a.buildAst();
    return a.buildAst();
  },
  color(_hash, ra, rb, ga, gb, ba, bb) {
    const val = [ra, rb, ga, gb, ba, bb].map((x) => x.sourceString).join("");
    return `#${val}`;
  },
  ValOp_absDiff(_a, _b, c, _d, _e, f) {
    throw new Error("absDiff not implemented");
    return {
      //   type: "ValOp_absDiff",
      a: c.buildAst(),
      b: f.buildAst(),
    };
  },
  ValOp_normal(left, op, right) {
    console.log("ValOp_normal");
    return {
      [op.sourceString]: {
        left: left.buildAst(),
        right: right.buildAst(),
      },
    };
  },

  number(digits) {
    return parseInt(digits.sourceString);
  },
  paren(_l, exp, _r) {
    return exp.buildAst();
  },
  ident(a) {
    console.log("iden", a.sourceString);
    return a.sourceString;
  },
  _terminal() {
    console.log("term");
    return null;
  },
  _iter(...children) {
    console.log("iter");
    return children.map((x) => x.buildAst());
  },
  NonemptyListOf(head, _, tail) {
    console.log("NonemptyListOf");
    return [head.buildAst(), ...tail.children.flatMap((c) => c.buildAst())];
  },
  letter(x) {
    console.log("letter");
    return x.sourceString;
  },
  Func1(name, _lParen, term1, _comma, term2, _rParen) {
    console.log("func 1");
    const funcPrep = funcTable[name.sourceString];
    if (!funcPrep) {
      throw new Error(`unknown function ${name.sourceString}`);
    }
    return funcPrep(term1.buildAst(), ...term2.buildAst());
  },
  Func2(name, _lParen, b, _comma, d, _e, f, _rParen) {
    const funcPrep = funcTable[name.sourceString];
    console.log("func 2");
    if (!funcPrep) {
      throw new Error(`unknown function ${name.sourceString}`);
    }
    return funcPrep(b.buildAst(), d.buildAst(), ...f.buildAst());
  },
};
semantics.addOperation("buildAst", actionDict);

const funcTable: Record<string, any> = {
  contrast: (left: string, right: string, algorithm: string) => ({
    contrast: { left, right },
    algorithm,
  }),
  cvdSim: (val: string, type: string) => ({ cvdSim: val, type }),
  name: (val: string) => ({ name: val }),
};

export default function compileToLL(input: string) {
  const match = gram.match(input, "Program");
  if (!match.succeeded()) {
    throw new Error("failed to parse");
  }
  return semantics(match).buildAst();
}
