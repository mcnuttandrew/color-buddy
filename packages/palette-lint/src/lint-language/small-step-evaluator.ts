import { Color, makePalFromString } from "color-buddy-palette";
import type { Palette } from "color-buddy-palette";
import { LLTypes, Environment } from "./lint-language";

type LLNode = InstanceType<(typeof LLTypes)["LLNode"]>;
export type InducedVariables = Record<string, Color | number>;

export function rewriteQuantifiers(node: any) {
  if (!node || !node.nodeType) {
    return node;
  }
  const newNode = copy(node);
  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      const left = rewriteQuantifiers(newNode.left);
      const right = rewriteQuantifiers(newNode.right);
      newNode.left = left;
      newNode.right = right;
      break;
    case "aggregate":
    case "array":
    case "conjunction":
    case "map":
      const children = node.children;
      if (Array.isArray(children)) {
        newNode.children = children.map(rewriteQuantifiers);
      } else {
        newNode.children = rewriteQuantifiers(children);
      }
      break;
    case "node":
    case "expression":
      newNode.value = rewriteQuantifiers(node.value);
      break;
    case "boolFunction":
    case "valueFunction":
      newNode.input = rewriteQuantifiers(node.input);
      break;
    case "quantifier":
      if (node.varbs.length === 1) {
        newNode.where = rewriteQuantifiers(node.where);
        newNode.predicate = rewriteQuantifiers(node.predicate);
      } else {
        const [head, ...tail] = node.varbs;
        const newSubNode = copy(node);
        newSubNode.where = rewriteQuantifiers(node.where);
        newSubNode.predicate = node.predicate;
        newSubNode.varbs = tail;
        newNode.where = undefined;
        newNode.varbs = [head];
        newNode.predicate = rewriteQuantifiers(newSubNode);
      }
      break;
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
    default:
      break;
  }
  return newNode;
}

function clearQuantifierResults(node: any) {
  if (!node) {
    return node;
  }
  if (!node.nodeType && !node.quant) {
    return node;
  } else if (node.quant) {
    return new LLTypes.LLBool(node.quantifierResult);
  }
  const newNode = copy(node);
  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      const left = clearQuantifierResults(newNode.left);
      const right = clearQuantifierResults(newNode.right);
      newNode.left = left;
      newNode.right = right;
      break;
    case "aggregate":
    case "array":
    case "conjunction":
    case "map":
      const children = node.children;
      if (Array.isArray(children)) {
        newNode.children = children.map(clearQuantifierResults);
      } else {
        newNode.children = clearQuantifierResults(children);
      }
      break;
    case "node":
    case "expression":
      newNode.value = clearQuantifierResults(node.value);
      break;
    case "valueFunction":
    case "boolFunction":
      newNode.input = clearQuantifierResults(node.input);
      break;
    case "quantifier":
      newNode.where = clearQuantifierResults(node.where);
      newNode.predicate = clearQuantifierResults(node.predicate);
      break;
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
    default:
      break;
  }
  return newNode;
}

function checkWhere(
  node: any,
  color: Color,
  varb: string,
  pal: Palette,
  inducedVariables: InducedVariables,
  index: number
): boolean {
  if (!node) return true;
  const result = evaluateNode(
    node,
    { ...inducedVariables, [varb]: color, [`index(${varb})`]: index + 1 },
    pal
  );
  return result.result;
}

function getValues(node: any, pal: Palette) {
  if (node?.input?.value === "colors") {
    return [...pal.colors];
  }
  if (
    Array.isArray(node?.input?.children) &&
    typeof node.input.children?.at(0)?.constructorString
  ) {
    return makePalFromString(
      node.input.children.map((x: any) => x.constructorString)
    ).colors;
  }

  return [];
}

const toVal = (x: Color | number) => {
  if (typeof x === "number") {
    return new LLTypes.LLValue(new LLTypes.LLNumber(x));
  } else if (typeof x === "string") {
    return new LLTypes.LLValue(
      new LLTypes.LLColor(Color.colorFromHex(x, "lab"), x)
    );
  }
  return new LLTypes.LLValue(new LLTypes.LLColor(x, x.toHex()));
};
const evalCache = new Map<string, any>();
function evaluateNode(
  node: any,
  inducedVariables: InducedVariables,
  pal: Palette
) {
  const serializedNode = JSON.stringify(node);
  const canUseCache = !serializedNode.includes("colors");
  const serializeKey = JSON.stringify({
    serializedNode,
    inducedVariables,
    background: pal.background.toHex(),
  });
  if (canUseCache && evalCache.has(serializeKey)) {
    return evalCache.get(serializeKey);
  }

  const opts = { debugParse: false, debugEval: false, debugCompare: false };

  const newEnv = Object.entries(inducedVariables).reduce(
    (acc, [key, value]) => acc.set(key, toVal(value)),
    new Environment(pal, {}, opts, {})
  );
  const cleanedNode = clearQuantifierResults(node);
  const result = cleanedNode.evaluate(newEnv);
  if (canUseCache) {
    evalCache.set(serializeKey, result);
  }
  return result;
}

function isValue(node: any) {
  switch (node.nodeType) {
    case "value":
      return true;
    case "expression":
    case "node":
      return isValue(node.value);
    case "bool":
    case "color":
    case "number":
    case "variable":
    case undefined:
      return true;
    case "array":
      return node.children.every(isValue);
    default:
      return false;
  }
}

function subTreeIsPureOp(
  node: any,
  inducedVariables: InducedVariables
): boolean {
  node.inducedVariables = toHexes(inducedVariables);
  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      return isValue(node.left) && isValue(node.right);
    case "aggregate":
    case "array":
    case "map":
      if (node.children?.nodeType === "variable") {
        return true;
      }
      const children = node.children;
      if (Array.isArray(children)) {
        return children.every((x: any) => isValue(x));
      } else {
        // for variables and such
        return isValue(children);
      }
    case "node":
    case "expression":
      return subTreeIsPureOp(node.value, inducedVariables);
    case "bool":
    case "color":
    case "number":
    case "value":
    case "variable":
      return false;
    case "boolFunction":
    case "valueFunction":
      return isValue(node.input);
    case "quantifier":
      return false;
    default:
      return false;
  }
}

function mapObject(obj: any, fn: (k: string, v: any) => any) {
  return Object.fromEntries(Object.entries(obj).map(([k, v]) => [k, fn(k, v)]));
}
function toHexes(obj: any) {
  return mapObject(obj, (k, v) => {
    if (typeof v === "number") {
      return v;
    }
    if (!v.toHex) {
      return v;
    }
    return v.toHex();
  });
}

function stripOutIds(str: string): string {
  return str.replace(/"id":\d+/g, '"id":0');
}

let counter = 0;
function traverseAndMaybeExecute(
  node: any,
  inducedVariables: InducedVariables,
  pal: Palette
): {
  result: any;
  didEval: boolean;
} {
  counter++;
  if (counter > 500) {
    throw new Error("Too many iterations");
  }
  const thisIsPureOp = subTreeIsPureOp(node, inducedVariables);
  const allChildrenPureOps = Array.isArray(node?.children)
    ? (node?.children || []).every((x: any) => isValue(x))
    : false;
  if (thisIsPureOp || allChildrenPureOps) {
    const preKey = JSON.stringify(node.copy());
    const result = evaluateNode(node, inducedVariables, pal).result;
    let astResult;
    if (Array.isArray(result)) {
      astResult = LLTypes.LLValueArray.tryToConstruct(result, {} as any);
    } else {
      astResult = LLTypes.LLValue.tryToConstruct(result, {} as any);
    }
    const postKey = JSON.stringify(astResult.copy());
    return {
      result: astResult,
      didEval: stripOutIds(preKey) !== stripOutIds(postKey),
    };
  }

  let updatedNode = copy(node);
  subTreeIsPureOp(updatedNode, inducedVariables);
  updatedNode.inducedVariables = toHexes({ ...inducedVariables });

  switch (node.nodeType) {
    case "pairFunction":
    case "numberOp":
    case "predicate":
      const leftTraverse = traverseAndMaybeExecute(
        updatedNode.left,
        inducedVariables,
        pal
      );
      if (leftTraverse.didEval) {
        updatedNode.left = leftTraverse.result;
        return { result: updatedNode, didEval: true };
      }
      const rightTraverse = traverseAndMaybeExecute(
        updatedNode.right,
        inducedVariables,
        pal
      );
      if (rightTraverse.didEval) {
        updatedNode.right = rightTraverse.result;
        return { result: updatedNode, didEval: true };
      }
      return { result: updatedNode, didEval: false };
    case "conjunction":
    case "array":
    case "aggregate":
    case "map":
      const children = updatedNode.children;
      if (Array.isArray(children)) {
        const newChildren = [];
        let found = false;
        for (let idx = 0; idx < children.length; idx++) {
          if (found) {
            newChildren.push(children[idx]);
            continue;
          }
          const child = children[idx];
          const childResult = traverseAndMaybeExecute(
            child,
            inducedVariables,
            pal
          );
          newChildren.push(childResult.result);
          found = childResult.didEval;
        }
        updatedNode.children = newChildren;
        return { result: updatedNode, didEval: found };
      } else {
        const childResult = traverseAndMaybeExecute(
          children,
          inducedVariables,
          pal
        );
        updatedNode.children = childResult.result;
        return { result: updatedNode, didEval: childResult.didEval };
      }
    case "node":
    case "expression":
      return traverseAndMaybeExecute(node.value, inducedVariables, pal);
    case "bool":
    case "color":
    case "number":
    case "value":
    case undefined:
      return { result: updatedNode, didEval: false };
    case "variable":
      return { result: updatedNode, didEval: false };
    case "boolFunction":
    case "valueFunction":
      const arg = traverseAndMaybeExecute(
        updatedNode.input,
        inducedVariables,
        pal
      );
      if (arg.didEval) {
        updatedNode.input = arg.result;
        return { result: updatedNode, didEval: true };
      } else {
        return { result: updatedNode, didEval: false };
      }

    case "quantifier":
      const values = getValues(updatedNode, pal);
      // const results = [];
      const results = values.map((color, i) => {
        const whereResult = checkWhere(
          updatedNode.where,
          color,
          updatedNode.varbs[0],
          pal,
          { ...updatedNode.inducedVariables },
          i
        );
        if (!whereResult) {
          return { result: "WHERE SKIP", didEval: true, color: color.toHex() };
        }
        const updatedVariables = {
          ...updatedNode.inducedVariables,
          [updatedNode.varbs[0]]: color,
          [`index(${node.varbs[0]})`]: i + 1,
        };
        const evals = smallStepEvaluator(
          updatedNode.predicate.value || updatedNode.predicate,
          { ...updatedVariables },
          pal
        );
        return {
          color: color.toHex(),
          evals: [...evals],
        };
      });
      const quantifierResult = evaluateNode(node, inducedVariables, pal).result;
      return {
        result: {
          results,
          quant: updatedNode.type,
          varb: updatedNode.varbs[0],
          quantifierResult,
        },
        didEval: true,
      };

    default:
      console.log(node.nodeType, " not implemented yet", node);
      throw new Error(`${node.nodeType} not implemented yet`, node);
    // return { result: node, didEval: false };
  }
}

function copy(node: any) {
  if (node.copy) {
    return node.copy();
  }
  return JSON.parse(JSON.stringify(node));
}

export function smallStepEvaluator(
  node: LLNode,
  inducedVariables: InducedVariables,
  pal: Palette,
  init?: boolean
): LLNode[] {
  const serializedNode = JSON.stringify(node);
  const canUseCache = !serializedNode.includes("colors");
  const serializeKey = JSON.stringify({
    serializedNode,
    inducedVariables,
    background: pal.background.toHex(),
    generate: "generate",
  });
  if (canUseCache && evalCache.has(serializeKey)) {
    console.log("cache hit");
    return evalCache.get(serializeKey);
  }
  if (init) {
    counter = 0;
  }
  let nodeCopy = copy(node);
  // weird hack to insert the induced variables over everything
  subTreeIsPureOp(nodeCopy, inducedVariables);
  const evalLog = node.nodeType === "quantifier" ? [] : [nodeCopy];
  let currentNode = copy(node);
  subTreeIsPureOp(node, inducedVariables);

  while (
    !subTreeIsPureOp(currentNode, { ...inducedVariables }) &&
    !isValue(currentNode)
  ) {
    const result = traverseAndMaybeExecute(
      currentNode,
      { ...inducedVariables },
      pal
    );
    result.result.inducedVariables = toHexes({ ...inducedVariables });

    evalLog.push(result.result);
    currentNode = result.result;
  }
  // get the final result
  const result = evaluateNode(node, { ...inducedVariables }, pal).result;
  const astResult = LLTypes.LLValue.tryToConstruct(result, {} as any);
  evalLog.push(astResult);
  // check if the last two items are the same, if so, remove the last one
  if (
    evalLog.length > 1 &&
    prepStr(evalLog[evalLog.length - 1]) ===
      prepStr(evalLog[evalLog.length - 2])
  ) {
    evalLog.pop();
  }
  if (canUseCache) {
    evalCache.set(serializeKey, evalLog);
  }
  return evalLog;
}

const prepStr = (node: any) => stripOutIds(JSON.stringify(copy(node)));
