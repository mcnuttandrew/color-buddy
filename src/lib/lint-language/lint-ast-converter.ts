// import type { SyntaxNodeRef } from "@lezer/common";
// import { parser } from "./lint-parser";
// import { Color } from "../Color";
// import {
//   LLNode,
//   LLExpression,
//   LLConjunction,
//   LLValueArray,
//   LLBool,
//   LLVariable,
//   LLColor,
//   LLNumber,
//   LLNumberOp,
//   LLPredicate,
//   LLValue,
//   LLValueFunction,
//   LLQuantifier,
//   LLReduces,
// } from "./lint-language";

// export function parseAST(programString: string): LLNode {
//   const ast = parser.parse(programString);
//   const result = traverseTree(ast.topNode, programString);
//   if (!result) throw new Error("Invalid AST");
//   return result;
// }

// function traverseTree(
//   node: SyntaxNodeRef,
//   programString: string
// ): LLNode | void {
//   const sliced = programString.slice(node.from, node.to);
//   // console.log(node.type.name, sliced);
//   let child = node.node.firstChild;
//   let children = [] as any[];
//   while (child) {
//     children.push(traverseTree(child, programString));
//     child = child.nextSibling;
//   }
//   children = children.filter((x) => (Array.isArray(x) ? x.length : x));
//   const symbols = new Set([
//     "AND",
//     "OR",
//     "NOT",
//     "LPAREN",
//     "RPAREN",
//     "SIMILAR",
//     "COUNT",
//     "SUM",
//     "MIN",
//     "MAX",
//     "MEAN",
//     "VARIABLE",
//     "COLOR",
//     "NUMBER",
//     "NUMBERCOMPARES",
//     "<",
//     ">",
//     "<=",
//     ">=",
//   ]);
//   if (symbols.has(node.type.name) || symbols.has(sliced) || sliced.length) {
//     return;
//   }
//   switch (node.type.name) {
//     case "true":
//     case "false":
//     case "BooleanExpression":
//       return new LLBool(sliced === "true");
//     case "AndExpression":
//       return new LLConjunction("and", children);
//     case "OrExpression":
//       return new LLConjunction("or", children);
//     case "NotExpression":
//       return new LLConjunction("not", children);
//     case "Identifier":
//       return new LLVariable(sliced);
//     case "ValueExpression":
//       return new LLValue(children[0]);
//     case "Program":
//       return children[0];
//     case "EqualExpression":
//       return new LLPredicate("==", children[0], children[1]);
//     case "NotEqualExpression":
//       return new LLPredicate("!=", children[0], children[1]);
//     case "GreaterThanExpression":
//       return new LLPredicate(">", children[0], children[1]);
//     case "LessThanExpression":
//       return new LLPredicate("<", children[0], children[1]);
//     case "SimilarExpression":
//       throw new Error("not done");
//       return new LLPredicate("similar", children[0], children[1]);
//     case "CountFunction":
//       return new LLReduces("count", children[0]);
//     case "SUM":
//       return new LLReduces("sum", children[0]);
//     case "MIN":
//       return new LLReduces("min", children[0]);
//     case "MAX":
//       return new LLReduces("max", children[0]);
//     case "MEAN":
//       return new LLReduces("mean", children[0]);
//     case "Variable":
//       return new LLVariable(sliced);
//     case "Color":
//       return new LLColor(Color.colorFromString(sliced));
//     case "Number":
//       return new LLNumber(+sliced);
//     case "NumberCompares":
//     // console.log("asd", sliced, children);
//     case "LParen":
//     case "RParen":
//       return;
//     case "âš ":
//       console.log("error?", sliced, children);
//       return;
//     default:
//       // if (!sliced.length || sliced === "(" || sliced === ")") {
//       //   return;
//       // }
//       console.log(node.type.name, sliced, children);
//       // throw new Error("Invalid node");
//       return;
//   }
// }
