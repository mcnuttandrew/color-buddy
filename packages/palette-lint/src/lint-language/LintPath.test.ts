import { expect, test } from "vitest";
import { GenerateAST } from "./lint-language";

const getAST = (node: any) => {
  const ast = (GenerateAST(node as any).value as any).children[0] as any;
  return ast;
};

test("Generating the eval paths works", () => {
  const exampleNode = {
    ">": {
      left: { deltaE: { left: "a", right: "b" }, algorithm: "2000" },
      right: 10,
    },
  };
  expect(getAST(exampleNode).generatePath([])).toMatchSnapshot();
});

test("Quantifier Test", () => {
  const quantifier = {
    all: {
      in: "colors",
      varb: "a",
      predicate: {
        ">": {
          left: {
            contrast: { left: "a", right: "background" },
            algorithm: "WCAG21",
          },
          right: 3,
        },
      },
      where: false,
    },
  };
  expect(getAST(quantifier).generatePath([])).toMatchSnapshot();
});
