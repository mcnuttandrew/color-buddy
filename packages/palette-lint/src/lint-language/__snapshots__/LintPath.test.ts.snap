// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generating the eval paths works 1`] = `
LLExpression {
  "id": 0,
  "nodeType": "expression",
  "path": [],
  "value": LLPredicate {
    "id": 0,
    "left": LLPairFunction {
      "id": 0,
      "left": LLVariable {
        "id": 0,
        "nodeType": "variable",
        "path": [
          ">",
          "left",
          "deltaE",
          "left",
        ],
        "value": "a",
      },
      "nodeType": "pairFunction",
      "params": {
        "algorithm": "2000",
      },
      "path": [
        ">",
        "left",
        "deltaE",
      ],
      "right": LLVariable {
        "id": 0,
        "nodeType": "variable",
        "path": [
          ">",
          "left",
          "deltaE",
          "right",
        ],
        "value": "b",
      },
      "type": "deltaE",
    },
    "nodeType": "predicate",
    "path": [
      ">",
    ],
    "right": LLNumber {
      "id": 0,
      "nodeType": "number",
      "path": [
        ">",
        "right",
      ],
      "value": 10,
    },
    "threshold": undefined,
    "type": ">",
  },
}
`;

exports[`Quantifier Test 1`] = `
LLExpression {
  "id": 0,
  "nodeType": "expression",
  "path": [],
  "value": LLQuantifier {
    "id": 0,
    "input": LLVariable {
      "id": 0,
      "nodeType": "variable",
      "path": [
        "all",
        "in",
      ],
      "value": "colors",
    },
    "nodeType": "quantifier",
    "path": [
      "all",
    ],
    "predicate": LLExpression {
      "id": 0,
      "nodeType": "expression",
      "path": [
        "all",
        "predicate",
      ],
      "value": LLPredicate {
        "id": 0,
        "left": LLPairFunction {
          "id": 0,
          "left": LLVariable {
            "id": 0,
            "nodeType": "variable",
            "path": [
              "all",
              "predicate",
              ">",
              "left",
              "contrast",
              "left",
            ],
            "value": "a",
          },
          "nodeType": "pairFunction",
          "params": {
            "algorithm": "WCAG21",
          },
          "path": [
            "all",
            "predicate",
            ">",
            "left",
            "contrast",
          ],
          "right": LLVariable {
            "id": 0,
            "nodeType": "variable",
            "path": [
              "all",
              "predicate",
              ">",
              "left",
              "contrast",
              "right",
            ],
            "value": "background",
          },
          "type": "contrast",
        },
        "nodeType": "predicate",
        "path": [
          "all",
          "predicate",
          ">",
        ],
        "right": LLNumber {
          "id": 0,
          "nodeType": "number",
          "path": [
            "all",
            "predicate",
            ">",
            "right",
          ],
          "value": 3,
        },
        "threshold": undefined,
        "type": ">",
      },
    },
    "type": "all",
    "varbs": [
      "a",
    ],
    "where": LLExpression {
      "id": 0,
      "nodeType": "expression",
      "path": [
        "all",
        "where",
      ],
      "value": LLBool {
        "id": 0,
        "nodeType": "bool",
        "path": [
          "all",
          "where",
        ],
        "value": false,
      },
    },
  },
}
`;
